vyper/parser/expr.py

```diff
@@ -521,117 +505,95 @@ def arithmetic(self):
+
             if ltyp == 'uint256' and isinstance(self.expr.op, ast.Add):
-                o = LLLnode.from_list([
-                    'seq',
-                    # Checks that: a + b >= a
-                    ['assert', ['ge', ['add', left, right], left]],
-                    ['add', left, right],
-                ], typ=BaseType('uint256', new_unit, new_positional), pos=getpos(self.expr))
+                # safeadd
+                arith = ['seq',
+                         ['assert', ['ge', ['add', 'l', 'r'], 'l']],
+                         ['add', 'l', 'r']]
+
             elif ltyp == 'uint256' and isinstance(self.expr.op, ast.Sub):
-                o = LLLnode.from_list([
-                    'seq',
-                    # Checks that: a >= b
-                    ['assert', ['ge', left, right]],
-                    ['sub', left, right]
-                ], typ=BaseType('uint256', new_unit, new_positional), pos=getpos(self.expr))
+                # safesub
+                arith = ['seq',
+                         ['assert', ['ge', 'l', 'r']],
+                         ['sub', 'l', 'r']]
+
             elif ltyp == rtyp:
-                o = LLLnode.from_list(
-                    [op, left, right],
-                    typ=BaseType(ltyp, new_unit, new_positional),
-                    pos=getpos(self.expr),
-                )
+                arith = [op, 'l', 'r']
+
             else:
                 raise Exception(f"Unsupported Operation '{op}({ltyp}, {rtyp})'")
+
         elif isinstance(self.expr.op, ast.Mult):
             if left.typ.positional or right.typ.positional:
                 raise TypeMismatchException("Cannot multiply positional values!", self.expr)
             new_unit = combine_units(left.typ.unit, right.typ.unit)
+            new_typ = BaseType(ltyp, new_unit)
+
             if ltyp == rtyp == 'uint256':
-                o = LLLnode.from_list([
-                    'if',
-                    ['eq', left, 0],
-                    [0],
-                    [
-                        'seq', ['assert', ['eq', ['div', ['mul', left, right], left], right]],
-                        ['mul', left, right]
-                    ],
-                ], typ=BaseType('uint256', new_unit), pos=getpos(self.expr))
+                arith = ['with', 'ans', ['mul', 'l', 'r'],
+                         ['seq',
+                             ['assert',
+                                 ['or',
+                                     ['eq', ['div', 'ans', 'l'], 'r'],
+                                     ['iszero', 'l']]],
+                             'ans']]
+
             elif ltyp == rtyp == 'int128':
-                o = LLLnode.from_list(
-                    ['mul', left, right],
-                    typ=BaseType('int128', new_unit),
-                    pos=getpos(self.expr),
-                )
+                # TODO should this be 'smul' (note edge cases in YP for smul)
```

Comment:
Let's fix this TODO

Reply:
ditto about the later clamp

Reply:
@fubuloubu I don't want to change semantics in this PR, TODO is a good flag for somebody to look at it later.